File: tianshu/core/rag/retrieve.py
from __future__ import annotations

import math
import re
from collections import Counter
from dataclasses import dataclass

from tianshu.core.rag.store import Chunk, all_chunks

_STOP = {
    "的", "了", "和", "是", "在", "有", "我", "你", "他", "她", "它", "这", "那",
    "与", "及", "或", "一个", "我们", "你们", "们", "也", "就", "都", "而", "但",
    "the", "a", "an", "is", "are", "was", "to", "of", "in", "for", "on", "and",
}


def _tokens(text: str) -> list[str]:
    text = text.lower()
    words = re.findall(r"[a-z0-9]+|[\u4e00-\u9fff]", text)
    return [w for w in words if w not in _STOP and (len(w) > 1 or "\u4e00" <= w <= "\u9fff")]


def _bm25_score(query_tokens: list[str], doc_tokens: Counter, df: dict[str, int], n_docs: int, avg_len: float, doc_len: int) -> float:
    k1, b = 1.5, 0.75
    score = 0.0
    for t in set(query_tokens):
        tf = doc_tokens.get(t, 0)
        if tf == 0:
            continue
        idf = math.log(1 + (n_docs - df.get(t, 0) + 0.5) / (df.get(t, 0) + 0.5))
        score += idf * (tf * (k1 + 1)) / (tf + k1 * (1 - b + b * doc_len / max(avg_len, 1)))
    return score


@dataclass
class Hit:
    chunk: Chunk
    score: float
    exact: bool


class RagRetriever:
    def __init__(self, chunks: list[Chunk] | None = None) -> None:
        self.chunks = chunks if chunks is not None else all_chunks()
        self._tokenized: list[Counter] = []
        self.df: Counter = Counter()
        total_len = 1
        for c in self.chunks:
            toks = Counter(_tokens(c.text))
            self._tokenized.append(toks)
            for t in toks:
                self.df[t] += 1
            total_len += len(c.text)
        self.n_docs = max(len(self.chunks), 1)
        self.avg_len = total_len / self.n_docs

    def search(self, query: str, top_k: int = 5) -> list[Hit]:
        tokens = _tokens(query)
        if not tokens or not self.chunks:
            return []
        scored = []
        for c, toks in zip(self.chunks, self._tokenized):
            score = _bm25_score(tokens, toks, self.df, self.n_docs, self.avg_len, len(c.text))
            if score > 0:
                exact = all(t in toks for t in tokens)
                scored.append(Hit(c, score, exact))
        scored.sort(key=lambda h: (h.exact, h.score), reverse=True)
        return scored[:top_k]

File: tianshu-ts/src/core/rag/retrieve.ts
import type { Chunk } from "./store.js";
import { allChunks } from "./store.js";

const STOP = new Set([
  "的", "了", "和", "是", "在", "有", "我", "你", "他", "她", "它", "这", "那",
  "与", "及", "或", "一个", "我们", "你们", "们", "也", "就", "都", "而", "但",
  "the", "a", "an", "is", "are", "was", "to", "of", "in", "for", "on", "and",
]);

function tokens(text: string): string[] {
  const t = text.toLowerCase();
  const words = t.match(/[a-z0-9]+|[\u4e00-\u9fff]/g) ?? [];
  return words.filter((w) => !STOP.has(w) && (w.length > 1 || (w >= "\u4e00" && w <= "\u9fff")));
}

function bm25Score(
  queryTokens: string[],
  docTokens: Map<string, number>,
  df: Map<string, number>,
  nDocs: number,
  avgLen: number,
  docLen: number,
): number {
  const k1 = 1.5;
  const b = 0.75;
  let score = 0.0;
  for (const t of new Set(queryTokens)) {
    const tf = docTokens.get(t) ?? 0;
    if (tf === 0) continue;
    const docFreq = df.get(t) ?? 0;
    const idf = Math.log(1 + (nDocs - docFreq + 0.5) / (docFreq + 0.5));
    score += idf * ((tf * (k1 + 1)) / (tf + k1 * (1 - b + (b * docLen) / Math.max(avgLen, 1))));
  }
  return score;
}

export interface Hit {
  chunk: Chunk;
  score: number;
  exact: boolean;
}

export class RagRetriever {
  chunks: Chunk[];
  private _tokenized: Array<Map<string, number>> = [];
  df = new Map<string, number>();
  nDocs: number;
  avgLen: number;

  constructor(chunks: Chunk[] | null = null) {
    this.chunks = chunks ?? allChunks();
    let totalLen = 1;
    for (const c of this.chunks) {
      const toks = new Map<string, number>();
      for (const t of tokens(c.text)) {
        toks.set(t, (toks.get(t) ?? 0) + 1);
      }
      this._tokenized.push(toks);
      for (const t of toks.keys()) {
        this.df.set(t, (this.df.get(t) ?? 0) + 1);
      }
      totalLen += c.text.length;
    }
    this.nDocs = Math.max(this.chunks.length, 1);
    this.avgLen = totalLen / this.nDocs;
  }

  search(query: string, topK = 5): Hit[] {
    const qTokens = tokens(query);
    if (!qTokens.length || !this.chunks.length) return [];
    const scored: Hit[] = [];
    this.chunks.forEach((c, i) => {
      const toks = this._tokenized[i];
      const score = bm25Score(qTokens, toks, this.df, this.nDocs, this.avgLen, c.text.length);
      if (score > 0) {
        const exact = qTokens.every((t) => (toks.get(t) ?? 0) > 0);
        scored.push({ chunk: c, score, exact });
      }
    });
    scored.sort((a, b) => Number(b.exact) - Number(a.exact) || b.score - a.score);
    return scored.slice(0, topK);
  }
}
